src/control/server/storage/scm/ipmctl_region.go

```diff
@@ -0,0 +1,376 @@
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return "", errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cmd := cmdShowRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return "", errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return out, nil
+}
+
+func (cr *cmdRunner) createRegions(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("set interleaved appdirect goal to create regions")
+
+	cmd := cmdCreateRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+func (cr *cmdRunner) removeRegions(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("set memory mode goal to remove regions")
+
+	cmd := cmdRemoveRegions
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+func (cr *cmdRunner) deleteGoals(sockID int) error {
+	if err := cr.checkIpmctl(badIpmctlVers); err != nil {
+		return errors.WithMessage(err, "checkIpmctl")
+	}
+
+	cr.log.Debug("delete any existing memory allocation goals")
+
+	cmd := cmdDeleteGoals
+	if sockID != sockAny {
+		cmd = fmt.Sprintf("%s --socket %d", cmd, sockID)
+	}
+
+	out, err := cr.runCmd(cmd)
+	if err != nil {
+		return errors.Wrapf(err, "cmd %q", cmd)
+	}
+	cr.log.Debugf("%q cmd returned: %q", cmd, out)
+
+	return nil
+}
+
+// socketRegionMap maps regions based on socket ID key.
+type socketRegionMap map[int]Region
+
+func (srm *socketRegionMap) keys() []int {
+	if srm == nil || len(*srm) == 0 {
+		return []int{}
+	}
+
+	keys := make([]int, 0, len(*srm))
+	for k := range *srm {
+		keys = append(keys, k)
+	}
+	sort.Ints(keys)
+
+	return keys
+}
+
+func (srm *socketRegionMap) isEmpty() bool {
+	if srm == nil || len(*srm) == 0 {
+		return true
+	}
+
+	return false
+}
+
+func (srm *socketRegionMap) fromXML(data string) error {
+	// parseRegions takes nvmxml output from ipmctl tool and returns PMem region details.
+	var rl RegionList
+	if err := xml.Unmarshal([]byte(data), &rl); err != nil {
+		return errors.Wrap(err, "parse show region cmd output")
+	}
+
+	regionsPerSocket := make(socketRegionMap)
+	for _, region := range rl.Regions {
+		sockID := int(region.SocketID)
+		if _, exists := regionsPerSocket[sockID]; exists {
+			return errors.Errorf("unexpected second region assigned to socket %d", sockID)
+		}
+		regionsPerSocket[sockID] = region
+	}
+	*srm = regionsPerSocket
```

Comment:
I don't think this is necessary, as maps in Go are [already reference types](https://go.dev/blog/maps). You should be able to use a non-pointer receiver for the method and just modify the map directly.

https://go.dev/play/p/FNtZyLnq4La

Reply:
done